### PR TITLE
SNOW-1019628 update file cleaner for snowpipe ingest

### PR DESCRIPTION
**Pull request by sfc-gh-gjachimko**
# Overview

SNOW-1019628

This PR addresses reported issue of cleaner not properly handling file cleanup - in some scenarios, files will be kept in staging area for as long as there is no message in the topic. 
This change modifies the file handling algorithm - it no longer depends on the first record to be received to start - it will start automatically for each partition. 

To enable this new cleaner, provide following configuration entry to your connector config:
`"snowflake.snowpipe.v2CleanerEnabled": true`

Core processing logic happens in ` StageFilesProcessor::trackFiles` 
This method is run in a worker thread. The only way to communicate with the cleaner, is through `ProgressRegister` interface, which allows adding new ingested files for tracking in a thread safe manner. 

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [x] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [ ] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [x] This change is protected by a config parameter  `snowflake.snowpipe.v2CleanerEnabled`.
    - [x] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [x] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java
Comment (sfc-gh-mbobowski) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:435]: `SnowflakeSinkTask#start()` -> `SnowflakeSinkServiceFactory.builder()` -> `SnowflakeSink ServiceV1`

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-lshcharbaty) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:405]: It isn't clear what `IngestEntry` really means. Let's name it `IngestedFileState` or somehow similar to it.

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-lshcharbaty) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:456]: Can we call `close` on `StageFilesProcessor` directly in `SnowflakeSinkServiceV1` and avoid having this awkward reference to an `owner` of this `ProgressRegisterImpl`?

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-lshcharbaty) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:464]: :+1:

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-lshcharbaty) [inline src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java:20]: Nit: wildcard import.

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-lshcharbaty) [inline src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java:14]: Nit: wildcard import.

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-lshcharbaty) [inline src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java:50]: I haven't seen anyone naming SUTs victims :)
It wasn't really obvious from the tests what is represented by it.

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-lshcharbaty) [inline src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java:71]: We can use awaitility instead.
```java
await().atMost(2, SECONDS).until(completed::get);
```

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-lshcharbaty) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:141]: Nit: we always pass `Thread.currentThread()` for the `worker` parameter.

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-lshcharbaty) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:108]: Do we really want to use a common `ForkJoinPool`? I assume it can be used by multiple different connectors (not only Kafka ones) if they're running in the same kafka connect instance.

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:8]: different IJ settings i guess .updated.

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java]: i don't think this code requries any dedicated scheduler TBH. common pool executor should do the job good enough.

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java]: ok, refactored.

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java]: first caller gets the reference and breaks the cleaner loop. all other will always get null and will not proceed. 
yes, it is ok to interrupt them - that's the best signal for the executing in methods like sleep or doing I/O to actually give up on execution and return control. 
i'm handling the interrupt exception in the trackFiles method, so in such case the interrupted flag would be cleared and thread returns to the pool as it would normally complete its job.

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-gjachimko) [reply src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java]: old habit :)

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-gjachimko) [reply src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java]: i didn't notice we have that library at hand !

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java]: and? i'd rather use the threads which are already there, present and ready for work, rather than spawn new ones. 
the moment thread from the pool is stuck on the I/O or timed wait state and pool is running out of threads - it will automatically add a new one. 
also - the lifetime of the sink service is controlled by kafka - it may go up and down quite frequently - in this case, not managing our own threads is even better idea.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionService.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionService.java:55]: nit: this sentence is reading weird!

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:96]: why is this set to null?

Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:96]: it never had a chance to be initialized to something non null - the object was constructed before this.telemetry is assigned - which is initialized later in the code. 
and i had to pass null explicit here, as i had to make the factory method static (to avoid code duplication)

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:37]: Thanks for javadocs but
minor nit:
can you use `<ul> <li>`, it will be more readable?

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:157]: should we use `ProgressRegister` interface here?

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:481]: nit: `ProgressRegisterImpl`

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:157]: implementation has some internal content - thus - the implementation here. the trackFiles itself should be considered private (shall we put it in dedicated package? - it would be effecitvely private for other callers)

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:481]: i didn't want to expose implementation details - what client gets - is a interface reference, thus the logger name here

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:157]: it will make testing difficult. but I dont have a major concern here.

**Review by sfc-gh-japatel — COMMENTED** [review]
lets discuss this in our meeting for coming friday. High level idea is good, i just wish the PR was a bit smaller. (Breaking up in smaller chunks would have made reviewer's life easy)

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionService.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionService.java:55]: extended

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:96]: ok, i've refactrored builder, so the telemetry would be accessible in the ctorl

File: src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java
Comment (sfc-gh-lshcharbaty) [inline src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java:176]: Nit: we don't really run the cleaner on the current thread.

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-lshcharbaty) [inline src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java:100]: Nit: it looks a little bit complex to grasp.
First we create a scheduler with `loops` passed as an argument. Then we alter the `loops` values so that during `victim.trackFiles` it's taken into account since we propagated it earlier.
Will it be simpler to have a method, let's say `createStageFilesProcessor`, that takes number of iterations as an argument? So we create it in each test without any indirection involving atomics?

File: src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java
Comment (sfc-gh-lshcharbaty) [inline src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java:130]: Nit: it's quite hard to read this test name. Adding underscores might help: `fileProcessor_willPurgeLoadedFiles_whenHistoryIsAvailableAfterFilesHaveBeenSubmitted`.
Applies to the other tests too.

**Review by sfc-gh-lshcharbaty — APPROVED** [review]

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java
Comment (sfc-gh-achyzy) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:81]: Why use setter in the constructor?

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java
Comment (sfc-gh-achyzy) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:335]: Please use a consistent way of null checking - I think in other places we use variable != null

File: src/test/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1Test.java
Comment (sfc-gh-achyzy) [inline src/test/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1Test.java:23]: Ideally we have our own fake (and abstraction over it SimpleIngestManager) instead of yet another mock.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java
Comment (sfc-gh-achyzy) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:340]: Why do you need more than 1 thread? It could impact the customer workload and since customers run multiple kafka connect clusters on the same node it could be too much to use the  availableProcessors

**Review by sfc-gh-achyzy — COMMENTED** [review]
Overall looks good. I'll wait for comment reply before the approval. 
 
For the future please split the refactoring and the core changes to seprate PRs (or at least provide more details list of going through the changes).

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1.java:81]: it was configured using builder - i've just reused the setter. i've updated the code to directly assign value.

File: src/test/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1Test.java
Comment (sfc-gh-gjachimko) [reply src/test/java/com/snowflake/kafka/connector/internal/SnowflakeIngestionServiceV1Test.java:23]: maybe, but there were none so far and I wanted to provide some basic testing for the service itself, without relying on the underlying SDK class

**Review by sfc-gh-achyzy — APPROVED** [review]
LGTM. 
Thanks.

**Pull request merged**
